cluster_experiments/synthetic_control_utils.py

```diff
@@ -0,0 +1,59 @@
+from functools import partial
+from itertools import product
+
+import numpy as np
+import pandas as pd
+from scipy.optimize import fmin_slsqp
+
+
+def loss_w(W: np.ndarray, X: np.ndarray, y: np.ndarray) -> float:
+    """
+    This function calculates the root mean square error (RMSE) between the actual and predicted values in a linear model.
+    It is used as an objective function for optimization problems where the goal is to minimize the RMSE.
+
+    Parameters:
+    W (numpy.ndarray): The weights vector used for predictions.
+    X (numpy.ndarray): The input data matrix.
+    y (numpy.ndarray): The actual output vector.
+
+    Returns:
+    float: The calculated RMSE.
+    """
+    return np.sqrt(np.mean((y - X.dot(W)) ** 2))
+
+
+def get_w(X, y, verbose=False) -> np.ndarray:
+    """
+    Get weights per unit, constraint in the loss function that sum equals 1; bounds 0 and 1)
+    """
+    w_start = np.full(X.shape[1], 1 / X.shape[1])
+    bounds = [(0.0, 1.0)] * len(w_start)
+
+    weights = fmin_slsqp(
+        partial(loss_w, X=X, y=y),
+        w_start,
+        f_eqcons=lambda x: np.sum(x) - 1,
+        bounds=bounds,
+        disp=verbose,
+    )
+    return weights
+
+
+def generate_data(N, start_date, end_date):
```

Comment:
why is this function needed in the library?

Reply:
I see it being used in a couple places:
* short example -> if it were only here, I'd add it examples/
* unit tests -> if it were only here, I'd add it tests/

since used in two it makes sense to keep it, but I'd make it internal, like `_generate_data`, and change the name to something more specific: `_generate_synthetic_control_data`

Reply:
I don't get it, if I keep it in utils and rename with underscore, then it would be wrong to import to other modules right? 

Do you mean to remove from modules and have 2 different definitions at short examples and unit tests?

Reply:
you're right, I don't have much control on the use of underscores, I was thinking adding it because it's not something that is useful to users (users should provide their data), but I agree that I'm skipping some standard rules when importing it. then never mind about underscore, just rename it to something like generate_synthetic_control_data